### PR TITLE
FMC-463: Fail task if error is encountered when writing to BigQuery

### PR DESCRIPTION
**Pull request by Cyril-Engels**
Background: 
The connector writes to BigQuery using a separate thread pool, which means that errors encountered during those writes don’t immediately cause the task to fail. Those errors aren’t swallowed silently, however, and do cause the task to throw an exception from its [`flush method`](https://github.com/wepay/kafka-connect-bigquery/blob/4fb38a8f4b3b6db065976639c57106db9896f0db/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java#L127) when invoking [`KCBQThreadPoolExecutor::awaitCurrentTasks`](https://github.com/wepay/kafka-connect-bigquery/blob/4fb38a8f4b3b6db065976639c57106db9896f0db/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java#L76-L100).

Unfortunately, the framework does not fail tasks when they throw exceptions from their [`flush method`](https://github.com/apache/kafka/blob/73ec7304b9e13cf4b9da05f215bd356a84efe0e7/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java#L384-L397) (which is indirectly called as a result of invoking SinkTask::preCommit), and instead, [logs the error and resets the consumer to the last successful offset](https://github.com/apache/kafka/blob/73ec7304b9e13cf4b9da05f215bd356a84efe0e7/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java#L384-L397).

We should do some digging into what the actual contract provided by the framework with regards to `SinkTask::flush` and `SinkTask::preCommit` is, specifically for when a call to them fails. It’s possible that this is a bug in the framework, but if the behavior we’re seeing is well-documented and intended, then the connector should be adjusted to throw the exception from somewhere else in order to actually fail the task.

This commit exposes the Exception in `BigQuerySinkTask::put`, so the task can fail without keeping reading the same batch of problematic records.

### CODE REVIEW

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java [hunk 1]
```diff
@@ -106,4 +106,16 @@ private static String createErrorString(Collection<Throwable> errors) {
...
   }
+
+  /**
+   * Checks for BigQuery errors. No-op if there isn't any error.
+   *
+   * @throws BigQueryConnectException if there is any error when writing to BigQuery.
+   */
+  public void checkForErrors() throws BigQueryConnectException {
+    if (encounteredErrors.size() > 0) {
+      throw new BigQueryConnectException("Encountered unrecoverable errors while writing to "
```
Comment (mageshn) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:117]: Would it make sense to include actual errors here?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/TableWriter.java [hunk 2]
```diff
@@ -85,6 +85,8 @@ public void run() {
           if (isBatchSizeError(err)) {
             failureCount++;
             currentBatchSize = getNewBatchSize(currentBatchSize);
+          } else {
+            throw err;
```
Comment (mageshn) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/TableWriter.java:89]: Why is this required?

Comment (Prathibha-m) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/TableWriter.java:89]: If we can also have a check for it not being a temporary network error (`backendError`) before throwing it, it might help

https://cloud.google.com/bigquery/docs/error-messages

Comment (Cyril-Engels) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/TableWriter.java]: It's already checked in `BigQueryWriter`.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java [hunk 3]
```diff
@@ -106,4 +106,18 @@ private static String createErrorString(Collection<Throwable> errors) {
...
     return String.join(", ", exceptionTypeStrings);
   }
+
+  /**
+   * Checks for BigQuery errors. No-op if there isn't any error.
+   *
+   * @throws BigQueryConnectException if there is any error when writing to BigQuery.
+   */
+  public void checkForErrors() throws BigQueryConnectException {
+    synchronized (encounteredErrors) {
```
Comment (mageshn) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:116]: I haven't had a chance to look at how this datastructure is used throughout. But just wanted to understand why we need to synchronize here. Is that absolutely necessary?

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTaskTest.java [hunk 4]
```diff
@@ -297,6 +297,36 @@ public void testPutWhenPartitioningOnMessageTimeWhenNoTimestampType() {
...
+
+    SchemaRetriever schemaRetriever = mock(SchemaRetriever.class);
+    SchemaManager schemaManager = mock(SchemaManager.class);
+
+    BigQuerySinkTask testTask = new BigQuerySinkTask(bigQuery, schemaRetriever, storage, schemaManager);
+    testTask.initialize(sinkTaskContext);
+    testTask.start(properties);
+
+    testTask.put(Collections.singletonList(spoofSinkRecord(topic)));
+    testTask.flush(Collections.emptyMap());
```
[the remark this answers is not in the document]
Comment (Cyril-Engels) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTaskTest.java:326]: A second call doesn't trigger exception.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java (hunk 3 above)
Comment (Cyril-Engels) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:116]: I think if we don't synchronize here, when we try to create error string later it might have already been cleared in `afterExecute`.

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTaskTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (C0urante) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTaskTest.java:326]: Shouldn't it, though? I thought the point of this PR was to trigger exceptions in calls to `put` instead of `flush`. 🤔

[the remark this answers is not in the document]
Comment (Cyril-Engels) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTaskTest.java:326]: Yes, but it doesn't happen immediately. Wait works.

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTaskTest.java [hunk 5]
```diff
@@ -297,6 +297,36 @@ public void testPutWhenPartitioningOnMessageTimeWhenNoTimestampType() {
...
+    SchemaRetriever schemaRetriever = mock(SchemaRetriever.class);
+    SchemaManager schemaManager = mock(SchemaManager.class);
+
+    BigQuerySinkTask testTask = new BigQuerySinkTask(bigQuery, schemaRetriever, storage, schemaManager);
+    testTask.initialize(sinkTaskContext);
+    testTask.start(properties);
+
+    testTask.put(Collections.singletonList(spoofSinkRecord(topic)));
+    Thread.sleep(100);
+    testTask.put(Collections.singletonList(spoofSinkRecord(topic)));
```
Comment (C0urante) [inline kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTaskTest.java:327]: It shouldn't be necessary to give it another record, right? Think we can just use:
```suggestion
    testTask.put(Collections.emptyList());
```

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTaskTest.java [hunk 6]
```diff
@@ -297,6 +297,36 @@ public void testPutWhenPartitioningOnMessageTimeWhenNoTimestampType() {
...
+
+    SchemaRetriever schemaRetriever = mock(SchemaRetriever.class);
+    SchemaManager schemaManager = mock(SchemaManager.class);
+
+    BigQuerySinkTask testTask = new BigQuerySinkTask(bigQuery, schemaRetriever, storage, schemaManager);
+    testTask.initialize(sinkTaskContext);
+    testTask.start(properties);
+
+    testTask.put(Collections.singletonList(spoofSinkRecord(topic)));
+    Thread.sleep(100);
```
Comment (C0urante) [inline kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTaskTest.java:326]: This is a little arbitrary and not guaranteed to be long enough for all environments (think: overworked Jenkins servers).

Could we use the [timeout](https://junit.org/junit4/javadoc/4.12/org/junit/Test.html#timeout()) option for the `@Test` annotation and put this in a loop, so that we'd basically keep retrying the operation every 100ms until either the timeout is exceeded or the expected failure occurs?

Thinking something like this:
```suggestion
while (true) {
  Thread.sleep(100);
  testTask.put(Collections.emptyList());
```

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java (hunk 3 above)
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:116]: Is it possible that a call to `put` causes writes to fail, then a call to `flush` causes `encounteredErrors` to get cleared (and throws an exception), and then the next call to `put` calls `checkForErrors` but since `encounteredErrors` has been cleared it doesn't cause an exception to be thrown?

Comment (Cyril-Engels) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:116]: I cannot eliminate this possibility. And that's why I put the error check after submitting tasks to the executor, because the insertion happens immediately, and why the wait solution is preferable. But I think this would be a very rare case.

Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:116]: I think if we know that the task is going to fail the next time that `put` is called, it might make sense to not clear `encounteredErrors` when it's checked in either `checkForErrors` or `awaitCurrentTasks`... is there a case when we'd ever want to clear that?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java [hunk 7]
```diff
@@ -185,6 +185,10 @@ private String getRowId(SinkRecord record) {
   @Override
   public void put(Collection<SinkRecord> records) {
     logger.info("Putting {} records in the sink.", records.size());
```
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:187]: Nit: might want to move this log statement after the error check, since we won't actually be putting anything in the sink if we're about to fail the task.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java (hunk 3 above)
Comment (mageshn) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:116]: what are the implications of not clearing in `awaitCurrentTasks`. Wouldn't we keep accumulating the same errors again and again? May be its not too bad.

Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:116]: I don't think so... we can consider a couple possible cases here:
- A call to `put` occurs and a resulting write fails, then a second call to `put` occurs and the task throws an exception and fails
- A call to `put` occurs and a resulting write fails, then `flush` is called and the task throws an exception, causing the offsets to rewind, then a second call to `put` occurs and the task throws an exception and fails
In neither of these cases will we see duplicate errors.

The only case I can think of where we would see duplicate errors would be this one:
- A call to `put` occurs and a resulting write fails, then `flush` is called and the task throws an exception, then _another_ call to `flush` occurs and the task throws the same exception all over again

This would be a pretty weird edge case, but even if it does happen, I think the behavior is somewhat reasonable. We wouldn't add new errors to `encounteredErrors` since that would only happen as the result of a call to `put`, which would throw an exception as soon as any of the writes fails with a non-retriable exception; we'd just end up throwing the same exception over and over again whenever `flush` got called until the framework finally invoked `put`.

Comment (mageshn) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java]: In scenario # 2 if we don't clear the error aren't we going to essentially get the same exception all over again? Its not a big deal but just wanted to call out the scenario.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java [hunk 8]
```diff
@@ -106,4 +105,16 @@ private static String createErrorString(Collection<Throwable> errors) {
                         .collect(Collectors.toList()));
     return String.join(", ", exceptionTypeStrings);
   }
+
+  /**
+   * Checks for BigQuery errors. No-op if there isn't any error.
+   *
+   * @throws BigQueryConnectException if there have been any unrecoverable errors when writing to BigQuery.
+   */
+  public void checkForErrors() throws BigQueryConnectException {
```
Comment (wicknicks) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:114]: maybe call this method `errorCount()` that returns the number of unrecoverable errors? and have the calling method throw an exception? right now, the signature seems a bit "bundled", where I am not really checking for anything. it blows up on touching, and there is no way to just "check", if that makes sense.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java (hunk 3 above)
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java]: Yeah, you'll potentially see the same exception in stack traces coming from both `flush` and `put`. We could consider changing the exception that gets thrown in `flush` altogether so that instead of being specific it just says something like "Cannot flush (or maybe commit offsets) because writes to BigQuery have failed", and then throwing the specific exception from `put`. Might make slogging through log messages a little easier for new users, but might make things harder for existing users who expect logs to take a certain format.

I think it's probably fine as-is 🤷

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java (hunk 8 above)
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:114]: I think an advantage of this design over a simple `errorCount` method is the ability to specify exactly what errors occurred when trying to write to BigQuery.

Some alternatives I can think of to make the API here clearer:

- Rename method to something like `throwIfFailed`, `maybeFail`, `maybeThrowWriteFails`, etc.
- Change method signature to either a `Throwable` that can be `null` if no writes have failed, or an `Optional<Throwable>` that contains a possible write failure

But TBH I don't mind what's present as-is. Will let you two settle this if it's a blocker

Comment (Cyril-Engels) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:114]: @wicknicks Since `encounteredErrors` can be modified concurrently, if we just return the number of errors to the caller, we can have the caller throw an exception, but we will lose the ability to include detailed error message. What do you think if we return an Optional of error string, and throw or no-op in the caller?

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/convert/BigQueryRecordConverterTest.java [hunk 9]
```diff
@@ -686,56 +686,25 @@ public void testInvalidMapSchemalessNestedMapNullValue() {
     Assert.assertEquals(kafkaConnectMap, stringObjectMap);
   }
 
-  @Test
-  public void testInvalidMapSchemalessNullValue() {
```
Comment (C0urante) [inline kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/convert/BigQueryRecordConverterTest.java:690]: Why are we removing these tests? Looks like they're still present in WePay's fork: https://github.com/wepay/kafka-connect-bigquery/blob/4fb38a8f4b3b6db065976639c57106db9896f0db/kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/convert/BigQueryRecordConverterTest.java#L658-L687

Comment (Cyril-Engels) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/convert/BigQueryRecordConverterTest.java:690]: No. It collided with the tests from upstream from last merge. We have duplicate tests in this file.

Comment (C0urante) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/convert/BigQueryRecordConverterTest.java:690]: Ahh, I see those now. Gotcha

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java (hunk 8 above)
Comment (wicknicks) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java:114]: yeah, this is a minor comment. ok to rename it also. any of @C0urante's suggestions are representative of what the method does (`maybeFail()` seems like the best)?
